File: ios/FoodHomeApp/Navigation/NavigationCoordinator.swift
import Foundation

/// Coordinates one trusted WKWebView generation without becoming an authorization boundary.
final class NavigationCoordinator {
    private let navigationPolicy: NavigationPolicy
    private let pendingDeepLinks: PendingDeepLinkStore
    private var navigate: ((URL) -> Void)?
    private var trustedDocumentReady = false
    private var currentAttachment: NavigationAttachment?
    private var currentDocument: TrustedDocumentAttachment?

    init(navigationPolicy: NavigationPolicy) {
        self.navigationPolicy = navigationPolicy
        self.pendingDeepLinks = PendingDeepLinkStore(navigationPolicy: navigationPolicy)
    }

    func classify(_ rawURL: String?) -> NavigationDecision {
        navigationPolicy.classify(rawURL)
    }

    func attach(navigate: @escaping (URL) -> Void) -> NavigationAttachment {
        let attachment = NavigationAttachment()
        currentAttachment = attachment
        currentDocument = nil
        self.navigate = navigate
        trustedDocumentReady = false
        return attachment
    }

    func detach(_ attachment: NavigationAttachment) {
        guard currentAttachment == attachment else { return }
        currentAttachment = nil
        currentDocument = nil
        navigate = nil
        trustedDocumentReady = false
    }

    @discardableResult
    func offer(_ rawURL: String) -> Bool {
        guard pendingDeepLinks.offer(rawURL) else { return false }
        deliverPendingRouteIfReady()
        return true
    }

    func markTrustedDocumentReady(
        _ attachment: NavigationAttachment,
        document: TrustedDocumentAttachment
    ) {
        guard currentAttachment == attachment, currentDocument == document else { return }
        trustedDocumentReady = true
        deliverPendingRouteIfReady()
    }

    func markTrustedDocumentLoading(
        _ attachment: NavigationAttachment
    ) -> TrustedDocumentAttachment? {
        guard currentAttachment == attachment else { return nil }
        let document = TrustedDocumentAttachment()
        currentDocument = document
        trustedDocumentReady = false
        return document
    }

    private func deliverPendingRouteIfReady() {
        guard let navigate,
              let route = pendingDeepLinks.consumeWhenReady(
                isBridgeReady: trustedDocumentReady
              )
        else {
            return
        }
        trustedDocumentReady = false
        currentDocument = nil
        navigate(route)
    }
}

struct NavigationAttachment: Equatable {
    fileprivate let id = UUID()
}

struct TrustedDocumentAttachment: Equatable {
    fileprivate let id = UUID()
}

File: android/app/src/main/java/market/foodhome/app/navigation/NavigationCoordinator.kt
package market.foodhome.app.navigation

/**
 * Coordinates one trusted WebView generation without becoming an authorization boundary.
 */
class NavigationCoordinator(
    private val navigationPolicy: NavigationPolicy,
) {
    private val pendingDeepLinks = PendingDeepLinkStore(navigationPolicy)
    private var navigate: ((String) -> Unit)? = null
    private var trustedDocumentReady = false
    private var nextAttachmentId = 0L
    private var nextDocumentId = 0L
    private var currentAttachment: NavigationAttachment? = null
    private var currentDocument: TrustedDocumentAttachment? = null

    @Synchronized
    fun classify(rawUrl: String?): NavigationDecision = navigationPolicy.classify(rawUrl)

    @Synchronized
    fun attach(navigate: (String) -> Unit): NavigationAttachment {
        val attachment = NavigationAttachment(++nextAttachmentId)
        currentAttachment = attachment
        currentDocument = null
        this.navigate = navigate
        trustedDocumentReady = false
        return attachment
    }

    @Synchronized
    fun detach(attachment: NavigationAttachment) {
        if (currentAttachment != attachment) return
        currentAttachment = null
        currentDocument = null
        navigate = null
        trustedDocumentReady = false
    }

    @Synchronized
    fun offerDeepLink(rawUrl: String): Boolean {
        if (!pendingDeepLinks.offer(rawUrl)) return false
        deliverPendingRouteIfReady()
        return true
    }

    @Synchronized
    fun markTrustedDocumentReady(
        attachment: NavigationAttachment,
        document: TrustedDocumentAttachment,
    ) {
        if (currentAttachment != attachment || currentDocument != document) return
        trustedDocumentReady = true
        deliverPendingRouteIfReady()
    }

    @Synchronized
    fun markTrustedDocumentLoading(
        attachment: NavigationAttachment,
    ): TrustedDocumentAttachment? {
        if (currentAttachment != attachment) return null
        val document = TrustedDocumentAttachment(++nextDocumentId)
        currentDocument = document
        trustedDocumentReady = false
        return document
    }

    private fun deliverPendingRouteIfReady() {
        val sink = navigate ?: return
        val route = pendingDeepLinks.consumeWhenReady(trustedDocumentReady) ?: return
        trustedDocumentReady = false
        currentDocument = null
        sink(route)
    }
}

data class NavigationAttachment internal constructor(internal val id: Long)
data class TrustedDocumentAttachment internal constructor(internal val id: Long)
